[Murugan Maniish] iP

Murugan-Maniish opened the pull request:
### Alastor

> "He-llo!" – Alastor ([Source](https://hazbinhotel.fandom.com/wiki/Alastor/Quotes))

Alastor frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/Murugan-Maniish/ip/releases/tag/v0.1)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😈 

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Alastor commands:
`list`
`mark` 
`unmark`
`todo`
`deadline` 
`event`
`find`
`bye`


If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
    public static void main(String[] args) {
        new Alastor("./data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

pikapika50 committed:
Level 0. Rename, Greet, Exit

pikapika50 committed:
Level 1. Echo

pikapika50 committed:
Level 2. Add, List

pikapika50 committed:
add Task.java

pikapika50 committed:
Level 3. Mark as Done

pikapika50 committed:
Level 3. Mark as Done (fixed - string split)

pikapika50 committed:
add ToDo.java, Deadline.java, Event.java

pikapika50 committed:
Level 4. ToDos, Events, Deadlines

pikapika50 committed:
A-TextUiTesting: Automated Text UI Testing

pikapika50 committed:
add DukeException.class

pikapika50 committed:
Level 5. Handle Errors (& write methods to reorganise code)

pikapika50 committed:
Level-6: Delete

pikapika50 committed:
Fix bug in Event.java which caused from and to to have the same values

pikapika50 committed:
Add Level-7: Save and Rewrite to abide by coding standards

For Level-7,
*Add initialise function to read tasks from file to chabot once its opened.
*Modify added function to enable appending of new tasks into file.
*Add editWrite function to replace the tasks in the file with those active in the list in the chatbot as it is hard to edit the file to mark/unmark/delete tasks.

For Coding Standards,
*Rewrite imports by replacing wildcard import with specific ones.
*As horizontalLine is a constant replace it with HORIZONTAL_LINE.

pikapika50 committed:
Update Exception Handling for initialise

Creates ./data dir and tasks.txt file if they are not present.

pikapika50 committed:
Add Level 8. Dates and Times

Updated Duke.java deadline and event function to create LocalDateTime object of by, from and to variables in a "dd-MM-yyyy HH:mm" from input date/time string. Added error handling for when date/time format in input string is incorrect.

Updated Deadline.java and Event.java to take LocalDateTime for variables by, from and to instead of String. Additionally updated their toString method to print out date/time in a different format from what was given by user as required.

pikapika50 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Update Exception Handling for initialise
  Add Level-7: Save and Rewrite to abide by coding standards

pikapika50 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add Level 8. Dates and Times

# Conflicts:
#	src/main/java/Duke.java

Resolution:
*fix import conflicts
*add error handling in initialise to deal with wrong date/time input, also enable skipping of lines so that it can read file without error
*add stringToDate function thats converts date/time String to LocalDateTime to avoid repeating code

pikapika50 committed:
Add toFile() function to Task, ToDo, Deadline and Event Classes

Add this function to make it easier to write to the file later on, instead of using additional arguments or instanceof.

pikapika50 committed:
Add minimal classes required for A-MoreOOP

Ui
*opens scanner and uses it to read commands from user in readCommand().
*All other methods are used to print responses to commands to user.

Storage
*load(): reads tasks from given file when chatbot is opened and returns a ArrayList<Task>.
*saveAdd(): adds tasks to file for todo/deadline/event commands.
*saveRewrite: rewrites entire file using TaskList for mark/unmark/delete commands.

TaskList
*encapsulates a ArrayList<Task> which can be initialised using two constuctors one which creates a new list and another that references a existing list.
*get(), size(), add() and delete() methods to use for commands.

Parser
*parseFile(): to read from File and add corresponding Task object to a TaskList.
*parseCommand(): to read commands from user and create corresponding Command Object.

pikapika50 committed:
Add *Command Classes and a abstract Command Class

Classes:
*Command {abstract}
*AddCommand for todo/deadline/event, takes in a Task object
*ListCommand for list
*MarkCommand for mark and unmark takes in a int index for task in list
*DeleteCommand for delete takes in a int index for task in list
*ExitCommand for bye
*InvalidCommand for non-existent commands throws exception

Each Class has 2 methods
*execute(): implements command using TaskList, Ui and Storage
*isExit(): tells whether it is a ExitCommand object

pikapika50 committed:
Update Duke.java for A-MoreOOP

pikapika50 committed:
Merge branch 'A-MoreOOP'

* A-MoreOOP:
  Update Duke.java for A-MoreOOP
  Add *Command Classes and a abstract Command Class
  Add minimal classes required for A-MoreOOP
  Add toFile() function to Task, ToDo, Deadline and Event Classes

sebin0817 commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    public void saveRewrite(TaskList tasks) throws DukeException {
```
For all public methods (except getters/setters and overriden methods), you should add Javadoc comments for clarity!

jmsandiegoo replied:
Javadocs would be great indeed! 👍

jmsandiegoo commented on src/main/java/AddCommand.java:
```diff
@@ -0,0 +1,20 @@
+public class AddCommand extends Command {
```
I like how you created classes for each related code logics such as for commands to ensure single responsibility principle. However, would it be better if the class would be in a package? I noticed it with the other classes as well.

jmsandiegoo replied:
I like how you created classes for each related code logics such as for commands to ensure single responsibility principle. However, would it be better if the class would be in a package? I noticed it with the other classes as well.

pikapika50 committed:
Organise into packages and Rename

Organise into a alastor package where all Command classes are in a sub-package called command and all Task classes are in a sub-package called task.

Rename "Duke" Duke.java and DukeException.java(and all instances of DukeException) to "Alastor" as Rename in level-0 was only done to strings.
Deleted Duke.java and DukeException.java to do so and created corresponding replacements. For unknown reason staged files show deletion of only DukeException of the two.
It also shows Command.java and InvalidCommand.java being deleted and new ones being created even though they were not deleted.

pikapika50 committed:
Merge branch 'branch-A-Packages'

* branch-A-Packages:
  Organise into packages and Rename

jmsandiegoo commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,4 @@
+public abstract class Command {
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException;
+    public abstract boolean isExit();
```
Could you consider adding one blank line between these two logical units?

jmsandiegoo commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,24 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
```
Aside from public methods, you can also try and include Javadoc header comments for your classes perhaps?

jmsandiegoo commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,87 @@
+                return new ExitCommand();
+            default:
+                return new InvalidCommand();
+        }
```
Should the case have the same indentation as the switch line for this statement?

jmsandiegoo reviewed (commented):
LGTM 💯, just some code style issues need to be fixed.

pikapika50 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

pikapika50 committed:
Override equals() for relevant Task Classes

This is done so that assertEquals can use it to check whether two Task objects are equal.
Did not need to do so for ToDo as the one in Task is sufficient but the other two types require overriding.

pikapika50 committed:
Update Parser.java

*Introduce enum ParseType to distinguish parsing file and parsing commands from using and giving corresponding error message.
*Add parameter to stringToDate() to account for ParseType.
*Rename checkForEmpty() to checkParameter() and Add parameter to account for ParseType. Additionally now it checks whether there is a missing or extra parameters.
*Update error messages
*Add checkParameter() in parseFile() to check whether file is corrupted. Also added Default in switch statement and check for marking value(either 0 or 1) in case file is corrupted.

pikapika50 committed:
Add ParserTest.java

Tests parserFile() function in 2 scenarios, one where the file in intact and the other where it is corrupted.

pikapika50 committed:
Add StorageTest.java

Tests load() method in 3 scenarios:
* where file exists
* where file does not exist
* where the directory the file is expected to be in (and the file) do not exist.

pikapika50 committed:
Merge branch 'branch-A-JUnit'

* branch-A-JUnit:
  Add StorageTest.java
  Add ParserTest.java
  Update Parser.java
  Override equals() for relevant Task Classes

pikapika50 committed:
Edit build.gradle for A-Jar

pikapika50 committed:
Tweak the code to comply with a coding standard

pikapika50 committed:
Add JavaDoc comments

pikapika50 committed:
Give users a way to find a task by searching for a keyword.

Made corresponding changes
*Parser.java: add case for find in parseCommand method to read find command form user
*TaskList.java: add find method to return matching list of tasks for Ui
*Ui.java: add showFind method to print out results for find using a TaskList (similar to list so maybe can simplify code)
*add FindCommand.java whose constructer takes in keyword for find
*Task.java: add isMatchingKeyword to check whether keyword can be found in task description

pikapika50 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Give users a way to find a task by searching for a keyword.

pikapika50 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments

# Conflicts:
#	src/main/java/alastor/TaskList.java
#	src/main/java/alastor/Ui.java

pikapika50 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak the code to comply with a coding standard

# Conflicts:
#	src/main/java/alastor/AlastorException.java
#	src/main/java/alastor/Parser.java

pikapika50 committed:
Use CheckStyle

pikapika50 committed:
Rectify CheckStyle errors

pikapika50 committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Rectify CheckStyle errors
  Use CheckStyle

pikapika50 committed:
Update build.gradle

pikapika50 committed:
Add images for user and chatbot

pikapika50 committed:
Update Command classes

They will return String when execute is called.

pikapika50 committed:
Update Ui to return String instead of print

pikapika50 committed:
Implement JavaFX

Followed tutorial.
Differences:
* Added greetings() in Alastor.java and called in Main.java to enable showing greeting in app.
* removed prefHeight in VBox in MainWindow.fxml to stop lists from being truncated.

pikapika50 committed:
Update Alastor.java

*change constructor to not have to arguments
*remove run() and main(String[] args)
*add getResponse(String input) and greeting() to enable app to show text/dialog

pikapika50 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Update Alastor.java
  Implement JavaFX
  Update Ui to return String instead of print
  Update Command classes
  Add images for user and chatbot
  Update build.gradle

pikapika50 committed:
Update build.gradle

So that jar file can create the application

pikapika50 committed:
Add JavaDoc for MainWindow.java showGreet method

pikapika50 committed:
Add assertions for A-Assertions

Let's,
*check that message in AlastorException is never null
*check that tasks are marked correctly in Storage.java load() method
*check line read from file is not null
*check command read from user is not null
*check task being written to file is not null
*check task list being rewritten to file is not null
*check that task is added/deleted from list in TaskList.java
*check that task marking are updated correctly in MarkCommand

pikapika50 committed:
Remove isExit() method from Command classes

The reason for this is because the method has become redundant.
The reason for this is that the exit command "bye" is dealt with in the
handleUserInput() method in MainWindow.java.

pikapika50 committed:
Remove isExit() method

pikapika50 committed:
Remove redundant methods and comment

The comment no longer serves any purpose.

For method in Ui.java
*showLine() is no longer need needed
*readCommand() is no longer used to read commands
*showError() is no longer used after aforementioned comment removed

These became redundant with the implementation of JavaFX.

pikapika50 committed:
Updated JavaDoc for getResponse method in Alastor.java

Murugan-Maniish committed:
Merge pull request #2 from Murugan-Maniish/branch-A-Assertions

Branch A-Assertions

pikapika50 committed:
Update so that user knows directory or file created

Before JavaFX was implemented, System.out.print() was used to tell
users when a directory or file was created but this has become
redundant now.

Two class attributes is added to Ui.java so that they can be updated
in Storage.java load() method so that the showGreet() method in
Ui.java can be modified to tell the user if any directory or file was
created. The two attributes will be made false again in the showGreet()
method so that it does not repeat the same message to the user even if
the relevant directory or file has been created previously and exists.

pikapika50 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertions for A-Assertions
  Add JavaDoc for MainWindow.java showGreet method

Murugan-Maniish committed:
Merge pull request #3 from Murugan-Maniish/branch-A-CodeQuality

Branch a code quality

pikapika50 committed:
Add help command

Let's
*create a HelpCommand class
*create a method in Ui to give a response
*let parser recognise help command

pikapika50 committed:
Merge branch 'branch-BCD-Extension'

* branch-BCD-Extension:
  Add help command

pikapika50 committed:
Improve the GUI

pikapika50 committed:
Give a unique personality

pikapika50 committed:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  Improve the GUI

pikapika50 committed:
Fix error in help response

pikapika50 committed:
Merge branch 'branch-A-Personality'

* branch-A-Personality:
  Fix error in help response
  Give a unique personality

pikapika50 committed:
More error handling

Data is not as expected:
*Start date/time is later (or same as) than end/time
*A value that should be unique is duplicated
 (e.g., two tasks with the same exact details)

pikapika50 committed:
Remove redundant images

pikapika50 committed:
Add a file named AI.md

pikapika50 committed:
Update AI.md

pikapika50 committed:
Improve response to user

*add spacing to make deadlines and events readable
*fix help response numbering

pikapika50 committed:
Ensure window shows Alastor as title

pikapika50 committed:
Add a representative screenshot

pikapika50 committed:
Add a User Guide

pikapika50 committed:
Remove redundant lines in ParserTest

pikapika50 committed:
Handle marking errors

pikapika50 committed:
Update README

pikapika50 committed:
Update README
